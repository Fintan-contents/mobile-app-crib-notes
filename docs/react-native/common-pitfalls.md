---
title: ハマりがちな落とし穴
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
