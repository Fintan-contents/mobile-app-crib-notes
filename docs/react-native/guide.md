---
title: 何らかのガイドを書きたい気持ちはある
---
