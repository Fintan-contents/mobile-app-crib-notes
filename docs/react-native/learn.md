---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。

ReactでWebアプリケーションを実装できるがReact Nativeでモバイルアプリケーションを実装したことがないアプリケーション開発者を対象として、次のことについて学ぶことを目的としています。

- ReactとReact Nativeの違い
- React Nativeで作成したモバイルアプリケーションの動作確認方法

また、作成したモバイルアプリケーションの配布方法やテスト手法についてはこのコンテンツでは触れません。

<!-- textlint-disable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->

import {PageList} from '../../src/components';
const overviews = [
  {
    title: 'Getting Started',
    to: '/react-native/learn/getting-started',
    summary: <>React Nativeでのアプリ開発に必要な環境を用意します。</>
  },
  {
    title: 'アプリの実装前に',
    to: '/react-native/learn/basic-concepts',
    summary: <>React Nativeでアプリを実装し始める前に最低限必要となる前提知識や、React NativeとReact Navigationの基本的な知識について学びます。</>
  },
  {
    title: 'ToDoアプリの実装',
    to: '/react-native/learn/todo-app',
    summary: <>React Nativeで簡単なToDoアプリを実装します。</>
  },
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length,ja-technical-writing/max-comma,ja-spacing/ja-no-space-around-parentheses,jtf-style/3.3.かっこ類と隣接する文字の間のスペースの有無,ja-technical-writing/ja-no-mixed-period,ja-technical-writing/no-unmatched-pair -->
