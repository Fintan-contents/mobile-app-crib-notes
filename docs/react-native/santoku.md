---
title: Example Application
sidebar_label: About
hide_table_of_contents: true
---

React Nativeサンプルアプリケーションの設計ドキュメントです。

<!-- textlint-disable ja-technical-writing/sentence-length -->

import {PageList} from '../../src/components';
const overviews = [
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length -->
