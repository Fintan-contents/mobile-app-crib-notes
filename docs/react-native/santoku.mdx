---
title: Example Application
sidebar_label: About
hide_table_of_contents: true
---

<!-- textlint-disable ja-technical-writing/sentence-length -->

import {PageList} from '../../src/components';
const overviews = [
  {
    title: 'Development',
    summary: '開発環境やビルドに関するドキュメント',
    to: '/react-native/santoku/development',
  }
]

<PageList overviews={overviews} colSize={12} />

<!-- textlint-enable ja-technical-writing/sentence-length -->
