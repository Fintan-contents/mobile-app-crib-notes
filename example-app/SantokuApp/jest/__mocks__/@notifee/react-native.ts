import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

// テストケースごとにモックは初期化しておく。
beforeEach(() =>
  Object.values(mock).forEach(fn => {
    if (jest.isMockFunction(fn)) {
      fn.mockClear();
    }
  }),
);

export default mock;
