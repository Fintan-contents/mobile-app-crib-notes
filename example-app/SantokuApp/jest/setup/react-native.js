// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
