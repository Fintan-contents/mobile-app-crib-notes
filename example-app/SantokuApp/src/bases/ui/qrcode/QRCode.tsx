import React from 'react';
import RNQRCode, {QRCodeProps as RNQRCodeProps} from 'react-native-qrcode-svg';

type QRCodeProps = RNQRCodeProps;

/**
 * モデル2のQRコードを生成し、表示するコンポーネント。
 * 指定されたデータや誤り訂正レベルに応じて以下の項目を自動計算します。
 * - エンコーディングモード
 * - シンボルバージョン
 * - マスキングパターン
 *
 * 誤り訂正レベルは、以下の4つのレベルから選択できます。（※括弧内の数値はデータが読み取れなくなる最大の損傷面を表します。）
 * デフォルト値は、"M"です。
 * - L（約7%）
 * - M（約15%）
 * - Q（約25%）
 * - H（約30%）
 *
 * 指定可能なPropsの詳細は、{@link https://github.com/awesomejerry/react-native-qrcode-svg#props}を参照してください。
 *
 * @see {@link https://github.com/awesomejerry/react-native-qrcode-svg}
 */
export const QRCode: React.FC<QRCodeProps> = (props: QRCodeProps) => {
  return <RNQRCode {...props} />;
};
