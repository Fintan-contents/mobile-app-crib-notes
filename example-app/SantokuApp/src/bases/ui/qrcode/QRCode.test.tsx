import {render, screen} from '@testing-library/react-native';
import beginnerMark from 'assets/images/beginner-mark.png';
import React from 'react';

import {QRCode} from './QRCode';

describe('QRCode', () => {
  it('QRCodeが正常にrenderできることを確認', () => {
    render(
      <QRCode
        value="0123456789"
        size={200}
        ecl="Q"
        backgroundColor="red"
        color="yellow"
        enableLinearGradient
        linearGradient={['rgb(255,255,255)', 'rgb(0,0,0)']}
        gradientDirection={['10', '10', '10', '10']}
        logo={beginnerMark}
        logoSize={50}
        logoBackgroundColor="white"
        logoMargin={4}
        logoBorderRadius={2}
        quietZone={5}
        onError={(e: unknown) => console.error(e)}
      />,
    );
    // QRCodeではtestIDを設定していないので、レンダリング後のコンポーネントを取得できずPropsの検証ができませんでした。
    // また、QRCode#getRefを使用して、SVGのデータURLを前回のテスト結果と比較することも検討しましたが、
    // render後のrefがundefinedになってしまい検証を諦めました。
    // https://github.com/awesomejerry/react-native-qrcode-svg#examples
    // そのため、スナップショットの検証のみを実施します。
    expect(screen).toMatchSnapshot();
  });
});
