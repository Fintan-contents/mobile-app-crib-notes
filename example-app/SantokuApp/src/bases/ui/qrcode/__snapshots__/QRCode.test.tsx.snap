// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRCode QRCodeが正常にrenderできることを確認 1`] = `
<RNSVGSvgView
  align="xMidYMid"
  bbHeight={200}
  bbWidth={200}
  focusable={false}
  height={200}
  meetOrSlice={0}
  minX={-5}
  minY={-5}
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      Object {
        "flex": 0,
        "height": 200,
        "width": 200,
      },
    ]
  }
  vbHeight={210}
  vbWidth={210}
  width={200}
>
  <RNSVGGroup>
    <RNSVGDefs>
      <RNSVGLinearGradient
        gradient={
          Array [
            0,
            -1,
            1,
            -16777216,
          ]
        }
        gradientTransform={null}
        gradientUnits={0}
        name="grad"
        x1="10"
        x2="10"
        y1="10"
        y2="10"
      />
    </RNSVGDefs>
    <RNSVGGroup>
      <RNSVGRect
        fill={4294901760}
        height={210}
        propList={
          Array [
            "fill",
          ]
        }
        width={210}
        x={-5}
        y={-5}
      />
    </RNSVGGroup>
    <RNSVGGroup>
      <RNSVGPath
        d="M0 4.761904761904762 L66.66666666666667 4.761904761904762 M76.19047619047619 4.761904761904762 L95.23809523809524 4.761904761904762 M104.76190476190476 4.761904761904762 L123.80952380952381 4.761904761904762 M133.33333333333334 4.761904761904762 L200 4.761904761904762 M0 14.285714285714285 L9.523809523809524 14.285714285714285 M57.14285714285714 14.285714285714285 L66.66666666666667 14.285714285714285 M76.19047619047619 14.285714285714285 L104.76190476190476 14.285714285714285 M133.33333333333334 14.285714285714285 L142.85714285714286 14.285714285714285 M190.47619047619048 14.285714285714285 L200 14.285714285714285 M0 23.80952380952381 L9.523809523809524 23.80952380952381 M19.047619047619047 23.80952380952381 L47.61904761904762 23.80952380952381 M57.14285714285714 23.80952380952381 L66.66666666666667 23.80952380952381 M76.19047619047619 23.80952380952381 L104.76190476190476 23.80952380952381 M133.33333333333334 23.80952380952381 L142.85714285714286 23.80952380952381 M152.38095238095238 23.80952380952381 L180.95238095238096 23.80952380952381 M190.47619047619048 23.80952380952381 L200 23.80952380952381 M0 33.33333333333333 L9.523809523809524 33.33333333333333 M19.047619047619047 33.33333333333333 L47.61904761904762 33.33333333333333 M57.14285714285714 33.33333333333333 L66.66666666666667 33.33333333333333 M76.19047619047619 33.33333333333333 L95.23809523809524 33.33333333333333 M114.28571428571428 33.33333333333333 L123.80952380952381 33.33333333333333 M133.33333333333334 33.33333333333333 L142.85714285714286 33.33333333333333 M152.38095238095238 33.33333333333333 L180.95238095238096 33.33333333333333 M190.47619047619048 33.33333333333333 L200 33.33333333333333 M0 42.857142857142854 L9.523809523809524 42.857142857142854 M19.047619047619047 42.857142857142854 L47.61904761904762 42.857142857142854 M57.14285714285714 42.857142857142854 L66.66666666666667 42.857142857142854 M76.19047619047619 42.857142857142854 L85.71428571428571 42.857142857142854 M95.23809523809524 42.857142857142854 L104.76190476190476 42.857142857142854 M114.28571428571428 42.857142857142854 L123.80952380952381 42.857142857142854 M133.33333333333334 42.857142857142854 L142.85714285714286 42.857142857142854 M152.38095238095238 42.857142857142854 L180.95238095238096 42.857142857142854 M190.47619047619048 42.857142857142854 L200 42.857142857142854 M0 52.38095238095238 L9.523809523809524 52.38095238095238 M57.14285714285714 52.38095238095238 L66.66666666666667 52.38095238095238 M95.23809523809524 52.38095238095238 L123.80952380952381 52.38095238095238 M133.33333333333334 52.38095238095238 L142.85714285714286 52.38095238095238 M190.47619047619048 52.38095238095238 L200 52.38095238095238 M0 61.9047619047619 L66.66666666666667 61.9047619047619 M76.19047619047619 61.9047619047619 L85.71428571428571 61.9047619047619 M95.23809523809524 61.9047619047619 L104.76190476190476 61.9047619047619 M114.28571428571428 61.9047619047619 L123.80952380952381 61.9047619047619 M133.33333333333334 61.9047619047619 L200 61.9047619047619 M76.19047619047619 71.42857142857143 L85.71428571428571 71.42857142857143 M95.23809523809524 71.42857142857143 L114.28571428571428 71.42857142857143 M9.523809523809524 80.95238095238095 L28.57142857142857 80.95238095238095 M38.095238095238095 80.95238095238095 L47.61904761904762 80.95238095238095 M57.14285714285714 80.95238095238095 L76.19047619047619 80.95238095238095 M85.71428571428571 80.95238095238095 L95.23809523809524 80.95238095238095 M104.76190476190476 80.95238095238095 L114.28571428571428 80.95238095238095 M133.33333333333334 80.95238095238095 L142.85714285714286 80.95238095238095 M152.38095238095238 80.95238095238095 L200 80.95238095238095 M0 90.47619047619047 L9.523809523809524 90.47619047619047 M38.095238095238095 90.47619047619047 L47.61904761904762 90.47619047619047 M95.23809523809524 90.47619047619047 L104.76190476190476 90.47619047619047 M114.28571428571428 90.47619047619047 L123.80952380952381 90.47619047619047 M133.33333333333334 90.47619047619047 L142.85714285714286 90.47619047619047 M180.95238095238096 90.47619047619047 L190.47619047619048 90.47619047619047 M0 100 L9.523809523809524 100 M57.14285714285714 100 L76.19047619047619 100 M85.71428571428571 100 L123.80952380952381 100 M152.38095238095238 100 L190.47619047619048 100 M0 109.52380952380952 L9.523809523809524 109.52380952380952 M28.57142857142857 109.52380952380952 L57.14285714285714 109.52380952380952 M95.23809523809524 109.52380952380952 L114.28571428571428 109.52380952380952 M133.33333333333334 109.52380952380952 L142.85714285714286 109.52380952380952 M180.95238095238096 109.52380952380952 L190.47619047619048 109.52380952380952 M19.047619047619047 119.04761904761904 L38.095238095238095 119.04761904761904 M47.61904761904762 119.04761904761904 L66.66666666666667 119.04761904761904 M76.19047619047619 119.04761904761904 L85.71428571428571 119.04761904761904 M133.33333333333334 119.04761904761904 L142.85714285714286 119.04761904761904 M152.38095238095238 119.04761904761904 L161.9047619047619 119.04761904761904 M190.47619047619048 119.04761904761904 L200 119.04761904761904 M76.19047619047619 128.57142857142858 L85.71428571428571 128.57142857142858 M95.23809523809524 128.57142857142858 L104.76190476190476 128.57142857142858 M123.80952380952381 128.57142857142858 L133.33333333333334 128.57142857142858 M142.85714285714286 128.57142857142858 L152.38095238095238 128.57142857142858 M180.95238095238096 128.57142857142858 L190.47619047619048 128.57142857142858 M0 138.0952380952381 L66.66666666666667 138.0952380952381 M76.19047619047619 138.0952380952381 L114.28571428571428 138.0952380952381 M123.80952380952381 138.0952380952381 L152.38095238095238 138.0952380952381 M0 147.61904761904762 L9.523809523809524 147.61904761904762 M57.14285714285714 147.61904761904762 L66.66666666666667 147.61904761904762 M85.71428571428571 147.61904761904762 L95.23809523809524 147.61904761904762 M123.80952380952381 147.61904761904762 L133.33333333333334 147.61904761904762 M142.85714285714286 147.61904761904762 L200 147.61904761904762 M0 157.14285714285714 L9.523809523809524 157.14285714285714 M19.047619047619047 157.14285714285714 L47.61904761904762 157.14285714285714 M57.14285714285714 157.14285714285714 L66.66666666666667 157.14285714285714 M76.19047619047619 157.14285714285714 L95.23809523809524 157.14285714285714 M104.76190476190476 157.14285714285714 L114.28571428571428 157.14285714285714 M123.80952380952381 157.14285714285714 L152.38095238095238 157.14285714285714 M190.47619047619048 157.14285714285714 L200 157.14285714285714 M0 166.66666666666666 L9.523809523809524 166.66666666666666 M19.047619047619047 166.66666666666666 L47.61904761904762 166.66666666666666 M57.14285714285714 166.66666666666666 L66.66666666666667 166.66666666666666 M85.71428571428571 166.66666666666666 L104.76190476190476 166.66666666666666 M133.33333333333334 166.66666666666666 L152.38095238095238 166.66666666666666 M180.95238095238096 166.66666666666666 L190.47619047619048 166.66666666666666 M0 176.19047619047618 L9.523809523809524 176.19047619047618 M19.047619047619047 176.19047619047618 L47.61904761904762 176.19047619047618 M57.14285714285714 176.19047619047618 L66.66666666666667 176.19047619047618 M76.19047619047619 176.19047619047618 L95.23809523809524 176.19047619047618 M114.28571428571428 176.19047619047618 L123.80952380952381 176.19047619047618 M152.38095238095238 176.19047619047618 L161.9047619047619 176.19047619047618 M190.47619047619048 176.19047619047618 L200 176.19047619047618 M0 185.71428571428572 L9.523809523809524 185.71428571428572 M57.14285714285714 185.71428571428572 L66.66666666666667 185.71428571428572 M76.19047619047619 185.71428571428572 L85.71428571428571 185.71428571428572 M95.23809523809524 185.71428571428572 L104.76190476190476 185.71428571428572 M133.33333333333334 185.71428571428572 L152.38095238095238 185.71428571428572 M161.9047619047619 185.71428571428572 L180.95238095238096 185.71428571428572 M0 195.23809523809524 L66.66666666666667 195.23809523809524 M95.23809523809524 195.23809523809524 L104.76190476190476 195.23809523809524 M114.28571428571428 195.23809523809524 L123.80952380952381 195.23809523809524 M133.33333333333334 195.23809523809524 L142.85714285714286 195.23809523809524 M152.38095238095238 195.23809523809524 L180.95238095238096 195.23809523809524 M190.47619047619048 195.23809523809524 L200 195.23809523809524 "
        propList={
          Array [
            "stroke",
            "strokeWidth",
            "strokeLinecap",
          ]
        }
        stroke={
          Array [
            1,
            "grad",
          ]
        }
        strokeLinecap={0}
        strokeWidth={9.523809523809524}
      />
    </RNSVGGroup>
    <RNSVGGroup
      matrix={
        Array [
          1,
          0,
          -0,
          1,
          71,
          71,
        ]
      }
    >
      <RNSVGDefs>
        <RNSVGClipPath
          name="clip-logo-background"
        >
          <RNSVGRect
            height={58}
            rx={2.16}
            ry={2.16}
            width={58}
            x={0}
            y={0}
          />
        </RNSVGClipPath>
        <RNSVGClipPath
          name="clip-logo"
        >
          <RNSVGRect
            height={50}
            rx={2}
            ry={2}
            width={50}
            x={0}
            y={0}
          />
        </RNSVGClipPath>
      </RNSVGDefs>
      <RNSVGGroup>
        <RNSVGRect
          clipPath="clip-logo-background"
          fill={4294967295}
          height={58}
          propList={
            Array [
              "fill",
            ]
          }
          width={58}
          x={0}
          y={0}
        />
      </RNSVGGroup>
      <RNSVGGroup
        matrix={
          Array [
            1,
            0,
            -0,
            1,
            4,
            4,
          ]
        }
      >
        <RNSVGImage
          align="xMidYMid"
          clipPath="clip-logo"
          height={50}
          meetOrSlice={1}
          src={null}
          width={50}
          x={0}
          y={0}
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGGroup>
</RNSVGSvgView>
`;
